domain/user/state/state.go

```diff
@@ -0,0 +1,335 @@
+`
+
+	_, err := tx.ExecContext(ctx, defineUserAuthenticationQuery, userUUID, false)
+	if err != nil {
+		return fmt.Errorf("cannot set authentification for user %q: %w", userUUID, err)
+	}
+
+	setActivationKeyQuery := `
+INSERT INTO user_activation_key (user_uuid, activation_key)
+VALUES (?, ?)
+ON CONFLICT(user_uuid) DO UPDATE SET activation_key = excluded.activation_key
+`
+
+	_, err = tx.ExecContext(ctx, setActivationKeyQuery, userUUID.String(), activationKey)
+	if err != nil {
+		return fmt.Errorf("cannot set activation key for user %q: %w", userUUID, err)
+	}
+
+	return nil
+}
+
+// AddUser will add a new user to the database. If the user already exists
+// an error that satisfies usererrors.AlreadyExists will be returned. If the
+// users creator is set and does not exist then an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUser(ctx context.Context, uuid domainuser.UUID, user user.User, creatorUUID domainuser.UUID) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		return addUser(ctx, tx, uuid, user, creatorUUID)
+	})
+
+	return err
+}
+
+// AddUserWithPasswordHash will add a new user to the database with the
+// provided password hash and salt. If the user already exists an error that
+// satisfies usererrors.AlreadyExists will be returned. If the users creator
+// does not exist or has been previously removed an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUserWithPasswordHash(
+	ctx context.Context,
+	uuid domainuser.UUID,
+	user user.User,
+	creatorUUID domainuser.UUID,
+	passwordHash string,
+	salt []byte,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err = addUser(ctx, tx, uuid, user, creatorUUID)
+		if err != nil {
+			return err
+		}
+
+		return setPasswordHash(ctx, tx, uuid, passwordHash, salt)
+	})
+
+	return err
+}
+
+// AddUserWithActivationKey will add a new user to the database with the
+// provided activation key. If the user already exists an error that
+// satisfies usererrors.AlreadyExists will be returned. if the users creator
+// does not exist or has been previously removed an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUserWithActivationKey(
+	ctx context.Context,
+	uuid domainuser.UUID,
+	user user.User,
+	creatorUUID domainuser.UUID,
+	activationKey string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err = addUser(ctx, tx, uuid, user, creatorUUID)
+		if err != nil {
+			return err
+		}
+		return setActivationKey(ctx, tx, uuid, activationKey)
+	})
+	if err != nil {
```

Comment:
This error handling is a bit out of whack for how the other functions are working. Can we align them all on a single approach?

Reply:
Look at the other Add* functions in state. They are returning the error directly from the transaction. I like this approach more by adding more context. Can we make them all consistent.